client/sftp_client.py

```diff
@@ -196,202 +614,749 @@ def load_dirs(self, path, parent=None):
+		self.signal_toggled_transfer_hidden = menu_item
+		menubar.append(menu_item)
+
+		menu_item = Gtk.SeparatorMenuItem()
+		menubar.append(menu_item)
+
+		menu_item = Gtk.MenuItem.new_with_label('Exit')
+		menu_item.connect('activate', self.shutdown)
+		menubar.append(menu_item)
+
+	def shutdown(self, _):
+		self.signal_window_destroy(None)
+		self.window.destroy()
+
+	def signal_toggled_transfer_hidden(self, _):
+		self.transfer_hidden = not self.transfer_hidden
+
+	def signal_toggled_validate_checksums(self, _):
+		self.validate = not self.validate
+
+	def _transfer_folder(self, task, ssh):
+		task.state = 'Transferring'
+		if task.parents is not None:
+			# prevents any lagging folders/files from causing issues
+			if  task.parents[0].state == 'Cancelled':
+				task.state = 'Cancelled'
+				return
+			if task.parents[0].state == 'Paused':
+				task.state = 'Paused'
+				return
+		if isinstance(task, UploadDirectoryTask):
+			new_dir = task.remote_path
+			dst = self.remote
+			if dst._already_exists_all(new_dir):
+				return
+			ftp = ssh.open_sftp()
+			dst._make_file(new_dir, ftp=ftp)
+			ftp.close()
+		elif isinstance(task, DownloadDirectoryTask):
+			new_dir = task.local_path
+			dst = self.local
+			if dst._already_exists_all(new_dir):
+				return
+			dst._make_file(new_dir)
+		if task.empty:
+			task.size = 0
+			task.transferred = 0
+			task.state = 'Completed'
+
+	def _transfer(self, task, ssh, chunk=0x1000):
+		task.state = 'Transferring'
+		ftp = ssh.open_sftp()
+		write_mode = 'ab+' if task.transferred > 0 else 'wb+'
+		try:
+			if isinstance(task, UploadTask):
+				task.size = self.local.get_file_size(task.local_path)
+				src_file_h = open(task.local_path, 'rb')
+				dst_file_h = ftp.file(task.remote_path, write_mode)
+			elif isinstance(task, DownloadTask):
+				task.size = self.remote.get_file_size(task.remote_path, stat_override=ftp.stat)
+				src_file_h = ftp.file(task.remote_path, 'rb')
+				dst_file_h = open(task.local_path, write_mode)
+			else:
+				raise ValueError('unsupported task type passed to _transfer')
+			src_file_h.seek(task.transferred)
+			while task.transferred < task.size:
+				if self._threads_shutdown.is_set():
+					task.state = 'Cancelled'
+				if task.state != 'Transferring':
+					if task.state == 'Cancelled':
+						if isinstance(task, UploadTask):
+							self.remote.remove_by_file_name(task.remote_path)
+						elif isinstance(task, DownloadTask):
+							self.local.remove_by_file_name(task.local_path)
+					break
+				temp = src_file_h.read(chunk)
+				dst_file_h.write(temp)
+				task.transferred += chunk
+			else:
+				task.state = 'Completed'
+				if self.validate:
+					src_file_h.seek(0)
+					dst_file_h.seek(0)
+					src_hash = hashlib.md5(src_file_h.read()).hexdigest()
+					dst_hash = hashlib.md5(dst_file_h.read()).hexdigest()
+					if src_hash == dst_hash:
+						self.application.logger.info("{0} and {1} have been validated".format(task.local_path, task.remote_path))
```

Comment:
Including the md5 checksum here in the logs would be nice. Also you can't use the `application.logger`, you need to make your own so the messages are correctly organized.